### PR TITLE
mod_itk: init at 2.4.7-04

### PR DESCRIPTION
**Pull request by NorfairKing**
###### Motivation for this change

I was asked to do this for @niteoweb and we figured we would contribute back the effort as well.

###### Things done
- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
  (I think.)

### CODE REVIEW

**Comment by NorfairKing** [comment]
cc @karantan who asked me to upstream this.

**Review by karantan — APPROVED** [review]
I've tested this on NixOS 21.05 and it worked.

You need to set `services.httpd.mpm = "prefork"; # mpm-itk needs "prefork"` and `extraModules = [ "proxy_fcgi" "status" { name = "mpm_itk"; path = "${pkgs.callPackage ./mod_itk.nix {}}/modules/mpm_itk.so"; } ];`.

After this gets merged you should be enable it like this: `extraModules = [ "proxy_fcgi" "status" pkgs.mod_itk ];`

**Comment by aanderse** [comment]
@karantan would it be better to add another entry to [services.httpd.mpm](https://github.com/NixOS/nixpkgs/blob/263ef4cc4146c9fab808085487438c625d4426a9/nixos/modules/services/web-servers/apache-httpd/default.nix#L584) and then update the module accordingly?

**Comment by karantan** [comment]
> @karantan would it be better to add another entry to [services.httpd.mpm](https://github.com/NixOS/nixpkgs/blob/263ef4cc4146c9fab808085487438c625d4426a9/nixos/modules/services/web-servers/apache-httpd/default.nix#L584) and then update the module accordingly?

From what I understand the `services.httpd.mpm` option defines how apache will handle multi-processing. It is true that`mpm-itk` is based on prefork MPM (i.e. non-threaded) but it's main purpose is to "allows you to run each of your vhost under a separate uid and gid". I would say it's more of a security module.

That said I'm not an apache expert so I could be wrong here.

**Comment by aanderse** [comment]
After reading through the documentation I'm fairly confident the "proper" way to do this would be what I said.

If you don't mind me asking what is your usecase for this mpm? Perhaps a bunch of sites running legacy cgi scripts?

**Comment by karantan** [comment]
> After reading through the documentation I'm fairly confident the "proper" way to do this would be what I said.

👍 

> If you don't mind me asking what is your usecase for this mpm? Perhaps a bunch of sites running legacy cgi scripts?

I was testing WordPress performance with apache & PHP-FPM vs apache & mod_itk. I had cases where PHP-FPM pools were consuming a lot of CPU and I wanted to test if anything changes if I remove it.

Technically I didn't need mod_itk for this, but it added the layer of security which PHP-FPM provided (i.e. stricter file and folder permissions).

**Comment by aanderse** [comment]
@karantan

`PHP-FPM` is very tweakable. When both `PHP-FPM` and `httpd` are tuned properly they should have superior performance to `mod_php` - and in theory a better footprint. Again, when configured properly the security is going to be far superior to `mod_itk` as well.

I can understand how `mod_itk` would be a reasonable choice for legacy `cgi` stuff, but my hunch is with a bit of work you're going to be more secure and performant, with a comparable to better footprint, than you would be with `mod_itk`.

I think this is a good PR to continue with regardless... but if you want a hand with your setup I could take a look, offer some ideas, and possibly get you in better shape.

Obviously I don't know all the details so I could be wrong, but let me know if you want a second set of eyes.

@NorfairKing did you want to go ahead and modify the `httpd` module to add `mod_itk` support in the `mpm` option?

**Comment by NorfairKing** [comment]
> @NorfairKing did you want to go ahead and modify the httpd module to add mod_itk support in the mpm option?

I just had a look at it but I'm afraid that my knowledge of this ecosystem isn't good enough to do this correctly.
`mod_itk` isn't named `mpm_something` so we can't just fit it into the `cfg.mpm` type as-is.
I would appreciate if you could fit it in the way you had in mind.

**Comment by aanderse** [comment]
> `mod_itk` isn't named `mpm_something` so we can't just fit it into the `cfg.mpm` type as-is.

The module name is misleading when you expect the `.so to match. This module in fact does fit :+1:
```
> tree /nix/store/4440p01sxyxxi52pvzh3kwnpb8z27drc-mod_itk-2.4.7-04                                                                                
/nix/store/4440p01sxyxxi52pvzh3kwnpb8z27drc-mod_itk-2.4.7-04
└── modules
    └── mpm_itk.so
```


> I would appreciate if you could fit it in the way you had in mind.

Definitely. Please integrate [this code](https://github.com/NixOS/nixpkgs/compare/master...aanderse:nixos/httpd-itk?expand=1) into your PR. @karantan - are you able to test please?

**Comment by NorfairKing** [comment]
@aanderse Thank you very much, I've integrated your change.

File: pkgs/servers/http/apache-modules/mod_itk/default.nix
Comment (aanderse) [inline pkgs/servers/http/apache-modules/mod_itk/default.nix:33]:
```suggestion
    description = "An MPM (Multi-Processing Module) for the Apache web server";
```

**Comment by karantan** [comment]
> Please integrate this code into your PR. @karantan - are you able to test please?

I'll try to test it today or tomorrow and I'll get back to you asap.

**Comment by karantan** [comment]
Hey, sorry for the late response. I had some problems rebuilding NixOS with these changes. I'm not sure what was the problem because I repeated the same process today and I was able to successfully rebuild this branch: https://github.com/NixOS/nixpkgs/compare/master...NorfairKing:mod_itk

But I think something is missing. Or maybe I'm just not configuring this correctly.

I added `mpm = "itk";` to the `services.httpd` section (rebuild nixos) and there was no changes to the /etc/httpd/httpd.conf file. I should see

```
    <IfModule itk.c>
        MaxClients           ${toString cfg.maxClients}
        MaxRequestsPerChild  ${toString cfg.maxRequestsPerChild}
    </IfModule>
```

but it's not there.

Also, I don't see that the module is loaded. I checked `ls /nix/store/x...-apache-httpd-2.4.47/modules/` and it's not there.

At this point I added "itk" to the `services.httpd.extraModules` list and repeated this process ^^ but the results were the same.

**Comment by aanderse** [comment]
@NorfairKing I spent some time on this tonight and found out something rather unfortunate... this apache module isn't a typical mpm like "prefork", "worker", and "event" are. This module actually **depends** on the "prefork" mpm, even though the documentation implies it is a replacement.

In conclusion, I was incorrect and you should drop the second commit. I apologize for having wasted your time on this. I think the upstream documentation isn't great :disappointed:

I'll merge this once we drop the second commit. Thanks for your patience and understanding.

**Comment by aanderse** [comment]
@karantan Once the second commit has been dropped I believe this is the configuration required to use this package effectively:
```nix
services.httpd.mpm = "prefork";
services.httpd.extraModules = [ { name = "mpm_itk"; path = "${pkgs.apacheHttpdModules.mod_itk}/modules/mpm_itk.so"; } ];
services.httpd.extraConfig = ''
  # put your mod_itk specific declarations here
'';

systemd.services.httpd.serviceConfig = {
  User = lib.mkForce "root";
  Group = lib.mkForce "root";
};
```

**Comment by NorfairKing** [comment]
Here we go, I've dropped the second commit.

**Review by aanderse — APPROVED** [review]
Thanks everyone!

**Comment by aanderse** [comment]
@ofborg eval

**Pull request merged**
